service/controller/release/resource/status/cluster.go

```diff
@@ -0,0 +1,205 @@
+
+	"github.com/giantswarm/release-operator/service/controller/key"
+)
+
+type TenantCluster struct {
+	ID               string
+	OperatorVersion  string
+	ProviderOperator string
+	ReleaseVersion   string
+}
+
+// Takes a list of tenant clusters and returns two maps containing the versions of their release and operator versions.
+func consolidateClusterVersions(clusters []TenantCluster) (map[string]bool, map[string]map[string]bool) {
+	releaseVersions := make(map[string]bool)
+	operatorVersions := make(map[string]map[string]bool)
+
+	// operatorVersions is a nested map including the operator name and version
+	// e.g. operatorVersions["aws-operator"]["8.7.6"]:true
+
+	for _, c := range clusters {
+		releaseVersions[c.ReleaseVersion] = true
+
+		if operatorVersions[c.ProviderOperator] == nil {
+			operatorVersions[c.ProviderOperator] = make(map[string]bool)
+		}
+		operatorVersions[c.ProviderOperator][c.OperatorVersion] = true
+	}
+
+	return releaseVersions, operatorVersions
+}
+
+// Returns a list of tenant clusters currently running on the installation.
+func (r *Resource) getCurrentTenantClusters(ctx context.Context) ([]TenantCluster, error) {
+
+	var tenantClusters []TenantCluster
+	{
+		awsClusters, err := r.getCurrentAWSClusters(ctx)
+		if IsResourceNotFound(err) {
+			// Fall through
+		} else if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		tenantClusters = append(tenantClusters, awsClusters...)
+
+		azureClusters, err := r.getCurrentAzureClusters(ctx)
+		if IsResourceNotFound(err) || IsNoMatchesForKind(err) {
+			// Fall through
+		} else if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		tenantClusters = append(tenantClusters, azureClusters...)
+
+		legacyClusters, err := r.getLegacyClusters(ctx)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		tenantClusters = append(tenantClusters, legacyClusters...)
+	}
+
+	return tenantClusters, nil
+}
+
+// Returns a list of AWS clusters according to the awscluster resource.
+func (r *Resource) getCurrentAWSClusters(ctx context.Context) ([]TenantCluster, error) {
+	awsclusters, err := r.k8sClient.G8sClient().InfrastructureV1alpha2().AWSClusters("default").List(ctx, metav1.ListOptions{})
+	if IsResourceNotFound(err) {
+		// Fall through
+	} else if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var clusters []TenantCluster
+	for _, cluster := range awsclusters.Items {
+		c := TenantCluster{
+			ID:               cluster.Name,
+			OperatorVersion:  cluster.Labels[apiexlabels.AWSOperatorVersion],
+			ReleaseVersion:   cluster.Labels[apiexlabels.ReleaseVersion],
+			ProviderOperator: key.ProviderOperatorAWS,
+		}
+		clusters = append(clusters, c)
+	}
+
+	return clusters, nil
+}
+
+// Returns a list of Azure clusters according to the azurecluster resource.
+func (r *Resource) getCurrentAzureClusters(ctx context.Context) ([]TenantCluster, error) {
+	azureClusters := azurecapi.AzureClusterList{}
+	err := r.k8sClient.CtrlClient().List(ctx, &azureClusters)
```

Comment:
Why does this use `CtrlClient` but `AWSCluster` uses `G8SClient`?

Reply:
AWS doesn't use upstream CAPI types, they come from apiextensions. Azure uses the upstream CAPZ types

Reply:
Got it.